src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolver.scala

```diff
@@ -0,0 +1,193 @@
+        case None =>
+          log.info("validation continue with binary searching")
+          requestBlockHeader(SearchState(minBlockNumber = 1, maxBlockNumber = bestBlockNumber, masterPeer = masterPeer))
+      }
+    case ResponseReceived(peer, BlockHeaders(blockHeaders), _) =>
+      log.warning(
+        "*** Invalid validation - Received {} block headers from a non-master peer {} ***",
+        blockHeaders.size,
+        peer
+      )
+    case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+    case Terminated(ref) => log.debug(s"TODO:... $ref") //TODO
+  }
+
+  private def binarySearchingLastValidBlock(searchState: SearchState, blockHeaderNumberToSearch: BigInt): Receive =
+    handleCommonMessages orElse {
+      case ResponseReceived(peer, BlockHeaders(Seq(childBlockHeader: BlockHeader)), timeTaken)
+          if peer == searchState.masterPeer && childBlockHeader.number == blockHeaderNumberToSearch =>
+        log.info("*** Received 1 block header searched in {} ms ***", timeTaken)
+        blockchain
+          .getBlockHeaderByNumber(childBlockHeader.number - 1)
+          .foreach(validateBlockHeaders(_, childBlockHeader, searchState))
+      case ResponseReceived(peer, BlockHeaders(blockHeaders), _) =>
+        log.warning("*** Invalid validation - Received {} block headers from peer {} ***", blockHeaders.size, peer)
+      case RequestFailed(peer, reason) => handleRequestFailure(peer, sender(), reason)
+      case Terminated(ref) => log.debug(s"TODO:... $ref") //TODO
+    }
+
+  private def validateBlockHeaders(
+      parentBlockHeader: BlockHeader,
+      childBlockHeader: BlockHeader,
+      searchState: SearchState
+  ): Unit = {
+    if (validateBlockHeaders(childBlockHeader, parentBlockHeader)) {
+      if (parentBlockHeader.number == searchState.minBlockNumber)
+        finalizeBranchResolver(blockchain.getBestBlockNumber(), parentBlockHeader.number, searchState.masterPeer)
+      else requestBlockHeader(searchState.copy(minBlockNumber = parentBlockHeader.number + 1))
+    } else if (parentBlockHeader.number == searchState.minBlockNumber)
+      finalizeBranchResolver(blockchain.getBestBlockNumber(), parentBlockHeader.number - 1, searchState.masterPeer)
+    else
+      requestBlockHeader(searchState.copy(maxBlockNumber = parentBlockHeader.number))
+  }
+
+  private def finalizeBranchResolver(lastBlock: BigInt, firstCommonBlockNumber: BigInt, masterPeer: Peer): Unit = {
+    discardLastInvalidBlocks(lastBlock, firstCommonBlockNumber)
+    context become receive
+    log.info(s"branch resolution completed - firstCommonBlockNumber $firstCommonBlockNumber")
+    fastSync ! BranchResolvedSuccessful(firstCommonBlockNumber = firstCommonBlockNumber, masterPeer = masterPeer)
+  }
+
+  private def getBestPeer: Option[(Peer, PeerInfo)] = {
+    handshakedPeers.toList.sortBy { case (_, peerInfo) => peerInfo.maxBlockNumber }(Ordering[BigInt].reverse).headOption
```

Comment:
I think we should perform an additional check here if this `bestPeer` is already blacklisted or not
I see this already addressed in https://jira.iohk.io/browse/ETCM-531